test/markup_renderer.go

```diff
@@ -0,0 +1,501 @@
+	attrs[name] = &value
+}
+
+func (r *markupRenderer) setResourceAttr(attrs map[string]*string, name string, rsc fyne.Resource) {
+	if rsc == nil {
+		return
+	}
+
+	if value := knownResource(rsc); value != "" {
+		r.setStringAttr(attrs, name, value)
+		return
+	}
+
+	var variant string
+	switch rsc.(type) {
+	case *theme.DisabledResource:
+		variant = "disabled"
+	case *theme.ErrorThemedResource:
+		variant = "error"
+	case *theme.InvertedThemedResource:
+		variant = "inverted"
+	case *theme.PrimaryThemedResource:
+		variant = "primary"
+	case *theme.ThemedResource:
+		variant = "default"
+	default:
+		r.setStringAttr(attrs, name, rsc.Name())
+		return
+	}
+
+	// That’s some magic to access the private `source` field of the themed resource.
+	v := reflect.ValueOf(rsc).Elem().Field(0)
+	src := reflect.NewAt(v.Type(), unsafe.Pointer(v.UnsafeAddr())).Elem().Interface().(fyne.Resource)
+	r.setResourceAttr(attrs, name, src)
+	r.setStringAttr(attrs, "themed", variant)
+}
+
+func (r *markupRenderer) setScaleModeAttr(attrs map[string]*string, name string, m canvas.ImageScale) {
+	var scaleMode string
+	switch m {
+	case canvas.ImageScaleSmooth:
+		// default mode, don’t add an attr
+	case canvas.ImageScalePixels:
+		scaleMode = "pixels"
+	default:
+		scaleMode = fmt.Sprintf("unknown scale mode: %d", m)
+	}
+	r.setStringAttr(attrs, name, scaleMode)
+}
+
+func (r *markupRenderer) setSizeAttr(attrs map[string]*string, name string, size fyne.Size) {
+	value := fmt.Sprintf("%dx%d", size.Width, size.Height)
+	attrs[name] = &value
+}
+
+func (r *markupRenderer) setStringAttr(attrs map[string]*string, name string, s string) {
+	if s == "" {
+		return
+	}
+	attrs[name] = &s
+}
+
+func (r *markupRenderer) writeCanvas(c fyne.Canvas) {
+	attrs := map[string]*string{}
+	r.setSizeAttr(attrs, "size", c.Size())
+	if tc, ok := c.(WindowlessCanvas); ok {
+		r.setBoolAttr(attrs, "padded", tc.Padded())
+	}
+	r.writeTag("canvas", false, attrs)
+	r.w.WriteRune('\n')
+	r.indentation++
+	r.writeTag("content", false, nil)
+	r.w.WriteRune('\n')
+	r.indentation++
+	driver.WalkVisibleObjectTree(c.Content(), r.writeCanvasObject, r.writeCloseCanvasObject)
+	r.indentation--
+	r.writeIndent()
+	r.writeCloseTag("content")
+	for _, o := range c.Overlays().List() {
+		r.writeTag("overlay", false, nil)
+		r.w.WriteRune('\n')
+		r.indentation++
+		driver.WalkVisibleObjectTree(o, r.writeCanvasObject, r.writeCloseCanvasObject)
+		r.indentation--
+		r.writeIndent()
+		r.writeCloseTag("overlay")
+	}
+	r.indentation--
+	r.writeIndent()
+	r.writeCloseTag("canvas")
+}
+
+func (r *markupRenderer) writeCanvasObject(obj fyne.CanvasObject, _, clipPos fyne.Position, clipSize fyne.Size) bool {
+	attrs := map[string]*string{}
+	r.setPosAttr(attrs, "pos", obj.Position())
+	r.setSizeAttr(attrs, "size", obj.Size())
+	if clipPos != fyne.NewPos(0, 0) || clipSize != fyne.NewSize(math.MaxInt32, math.MaxInt32) {
+		r.setRectAttr(attrs, "clip", clipPos, clipSize)
```

Comment:
I don't think this should be rendered - it is a property of the driver / renderer not the widget.
Also it gets in the way of this markup being writable manually...